Gems/O3DE/GeomNodes/Code/Source/Editor/Rendering/GNMeshController.cpp

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Editor/Rendering/GNMeshController.h>
+#include <Editor/Rendering/GNRenderMesh.h>
+#include <Editor/UI/Utils.h>
+#include <Editor/Common/GNConstants.h>
+#include <Editor/EBus/EditorGeomNodesComponentBus.h>
+
+#include <AzCore/Debug/Profiler.h>
+#include <AzCore/Component/NonUniformScaleBus.h>
+#include <AzCore/std/string/regex.h>
+#include <AzCore/Utils/Utils.h>
+#include <AzToolsFramework/API/ToolsApplicationAPI.h>
+#include <AzToolsFramework/API/EntityCompositionRequestBus.h>
+#include <AzToolsFramework/ToolsComponents/TransformComponent.h>
+#include <AzToolsFramework/Entity/EditorEntityHelpers.h>
+#include <AtomLyIntegration/CommonFeatures/Mesh/MeshComponentConstants.h>
+#include <AtomLyIntegration/CommonFeatures/Material/MaterialComponentConstants.h>
+#include <AtomLyIntegration/CommonFeatures/Mesh/MeshComponentBus.h>
+
+namespace GeomNodes
+{
+    GNMeshController::GNMeshController(AZ::EntityId entityId)
+        : m_entityId(entityId)
+    {
+        if (!m_renderMesh)
+        {
+            m_renderMesh = AZStd::make_unique<GNRenderMesh>(m_entityId);
+        }
+
+		AZ::TransformNotificationBus::Handler::BusConnect(m_entityId);
+		AzFramework::AssetCatalogEventBus::Handler::BusConnect();
+
+		m_world = AZ::Transform::CreateIdentity();
+		AZ::TransformBus::EventResult(m_world, m_entityId, &AZ::TransformBus::Events::GetWorldTM);
+    }
+
+    GNMeshController::~GNMeshController()
+    {
+		AzFramework::AssetCatalogEventBus::Handler::BusDisconnect();
+		AzToolsFramework::EditorComponentSelectionRequestsBus::Handler::BusDisconnect();
+		AzFramework::BoundsRequestBus::Handler::BusDisconnect();
+		AZ::TransformNotificationBus::Handler::BusDisconnect();
+		m_renderMesh.reset();
+    }
+
+    AZ::Aabb GNMeshController::GetEditorSelectionBoundsViewport(const AzFramework::ViewportInfo& /*viewportInfo*/)
+    {
+        return GetWorldBounds();
+    }
+    bool GNMeshController::EditorSelectionIntersectRayViewport(const AzFramework::ViewportInfo& /*viewportInfo*/, const AZ::Vector3& src, const AZ::Vector3& dir, float& distance)
```

Comment:
nit: some functions are missing a space between them

Reply:
noted. btw on viewportInfo parameter it might be better to use [[maybe_unused]] attribute for consistency?

Reply:
fixed in commit [d756b38](https://github.com/o3de/o3de-technicalart/pull/11/commits/d756b38541172fd3e0051140ca96cc9bd8b05388)